src/Connection.php

```diff
@@ -1816,6 +1817,11 @@ private function handleDriverException(
             $this->close();
         }
 
+        if ($exception instanceof DeadlockException) {
+            //Reset transaction nesting level since deadlocks always rollback.
```

Comment:
This should be true for all supported platforms and ideally proven by an integration test. At this point, it sounds like instead of tracking the transaction level within the DBAL itself, if possible, the DBAL should delegate it to the driver (e.g. [`PDO::inTransaction()`](https://www.php.net/manual/en/pdo.intransaction.php)).

Reply:
I've added a single test to validate that the transaction level is properly reset.

As far as I can tell moving this logic inside the driver would not be possible without major refactoring.

Also mysql doesn't have a way for telling if there is currently presented transaction.

The `PDO:inTransaction()` seems to do what actually DBAL does and monitors the the calls made by the driver's functions.

side note, took a step further and introduced new exception classes to allow application developers distinguish different cases without the need of comparing strings. (there is no active transaction, unable to alter nested transaction behavior, etc).

This should allow one to monitor "no active transaction is present" in the rollback phase and simply ignore the error if happens after `deadlock`. (Probably the real fix is to stop checking if there is active transaction when calling rollback?).